Python/Product/PythonTools/PythonTools/Commands/SendToReplCommand.cs

```diff
@@ -45,26 +56,76 @@ class SendToReplCommand : Command {
             var activeView = CommonPackage.GetActiveTextView(_serviceProvider);
             var project = activeView.GetProjectAtCaret(_serviceProvider);
             var analyzer = activeView.GetAnalyzerAtCaret(_serviceProvider);
+            ITextSelection selection = activeView.Selection;
+            ITextSnapshot snapshot = activeView.TextBuffer.CurrentSnapshot;
 
             var repl = ExecuteInReplCommand.EnsureReplWindow(_serviceProvider, analyzer, project);
-            repl.Show(true);
 
-            var version = analyzer.InterpreterFactory.Configuration.Version.ToLanguageVersion();
+            string input;
+            bool focusRepl = false;
+
+            if (selection.StreamSelectionSpan.Length > 0) {
+                // Easy, just send the selection to the interactive window.
+                input = activeView.Selection.StreamSelectionSpan.GetText();
+                if (!input.EndsWith("\n")) {
+                    input += activeView.Options.GetNewLineCharacter();
+                }
+                focusRepl = true;
+            } else if (!activeView.Properties.ContainsProperty(_executedLastLine)) {
+                // No selection, and we haven't hit the end of the file in line-by-line mode.
+                // Send the current line, and then move the caret to the next non-blank line.
+                ITextSnapshotLine targetLine = snapshot.GetLineFromPosition(selection.Start.Position);
+                input = targetLine.GetText();
+
+                bool moved = false;
+                while (targetLine.LineNumber < snapshot.LineCount - 1) {
+                    targetLine = snapshot.GetLineFromLineNumber(targetLine.LineNumber + 1);
+                    // skip over blank lines, unless it's the last line, in which case we want to land on it no matter what
+                    if (!string.IsNullOrWhiteSpace(targetLine.GetText()) ||
+                        targetLine.LineNumber == snapshot.LineCount - 1) {
+                        activeView.Caret.MoveTo(new SnapshotPoint(snapshot, targetLine.Start));
+                        activeView.Caret.EnsureVisible();
+                        moved = true;
+                        break;
+                    }
+                }
 
-            repl.Show(true);
+                if (!moved) {
+                    // There's no where for the caret to go, don't execute the line if 
+                    // we've already executed it.
+                    activeView.Caret.PositionChanged += Caret_PositionChanged;
+                    activeView.Properties[_executedLastLine] = _executedLastLine;
+                }
+            } else if (repl.InteractiveWindow.CurrentLanguageBuffer.CurrentSnapshot.Length != 0) {
```

Comment:
`CurrentLanguageBuffer` can be null at weird times - I'd rather not crash because of it. `(repl.InteractiveWindow.CurrentLanguageBuffer?.CurrentSnapshot.Length ?? 0) != 0` should be good enough handling here, and we can Assert in the else block below.